src/Lecture2.hs

```diff
@@ -264,7 +397,20 @@ data EvalError
 It returns either a successful evaluation result or an error.
 -}
 eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
+eval vars expr = case expr of
+     Lit value -> Right value
+     Var varName ->
+        let varValue = lookup varName vars
+        in case varValue of
+          Nothing    -> Left (VariableNotFound varName)
+          Just value -> Right value
+     {- I am sure there is more elegant solution, sth like Reactor zip / Mutiny combine -}
+     Add expr1 expr2 ->
+        let res1 = eval vars expr1
+            res2 = eval vars expr2
+        in case res1 of
+           Right r1      -> fmap (\r2 -> r2 + r1) res2
+           Left error    -> Left error
```

Comment:
That's a great helper function! 👏🏻 
Once you learn about Monads, you'll be able to reduce this function to a single line ✂️